### DISCUSSION TITLE
external_call_dll

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#30] [Announcements]
# 调用动态链接库

https://lv.qizhen.xyz/external_call_dll

**Comment by Arita9527** [comment]
你好，遇到DLL难题了。
1.导入马达控制DLL时，使用的是绝对路径，编程模式下马达运行没有问题。
2.但是我生成EXE之后，用EXE运行时，马达无法运行。
![调试时](https://i.postimg.cc/SKY9Y6jd/DLL-edit.png)

问题：
1。绝对路径DLL后面带个星号是什么意思？需要消除吗
2.为什么EXE之后不能动了呢？是不是发布时候要选择一些配置？
3.最优解决办法是什么呢？
谢谢

**Reply by Arita9527** [reply]
那种方法比较好？
1.直接把D盘的DLL 相关VI（写好的功能模块VI包含了dll小VI）拖进框图
2.先去框图里导入DLL到共享库，然后从控件面板把dll小VI一个一个的拖进来。再重新写一个功能模块。

**Reply by ruanqizhen** [reply]
这一部分我记不太清楚了，而且我用的免费版LabVIEW也没法测试这一部分。我只能提供一些想法：
我看你选了“Specify path on diagram”，这表明路径跟你在程序中的一些设定相关，可能会引起问题。你把它去掉试试看。
另外，使用相对路径或者系统路径会更灵活，因为如果你的程序还要运行在其它电脑上，DLL绝对路径可能会不同。不知道你的DLL是否可以被放在任何路径下运行，如果可以，可以把它和VI或者EXE都放在一起，同一个目录下。这样容易被找到。

还有你说的两种方法，没太大差别，全看你的需要。如果你已经有写好的功能模块VI了，就没必要再导入一次了。

**Reply by Arita9527** [reply]
谢谢
我百度了一下。有个博主解决方法
https://blog.csdn.net/sinat_33370970/article/details/110178911
我试了所有的选项，都不行。。。
不过他在文章里面提到”与调用方相同“。他们几个选项有什么不一样啊？
![](https://img-blog.csdnimg.cn/20201126111515742.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L3NpbmF0XzMzMzcwOTcw,size_16,color_FFFFFF,t_70)

**Reply by ruanqizhen** [reply]
这个因为我没法测试（免费版的LabVIEW没这个功能），所以也不太敢肯定的说”与调用方相同“是什么意思。只能你自己试试看了。
我的建议是尽量把DLL都放在LabVIEW的搜索路径下，比如系统路径下，这样问题可能会少一些。

**Reply by Arita9527** [reply]
好的。谢谢。
对了
绝对路径中有时候我看到IMCnet.DLL，有时候我看到IMCnet.*
这个星号是什么意思呢？

**Reply by ruanqizhen** [reply]
这个应该是跨平台用的，Windows下的动态链接库是 .dll文件，但在Linux 下动态链接库是 .a 文件。.* 表示在不同平台都能使用。如果你的程序只支持Windows，用dll就行。

**Comment by ruanqizhen** [comment]
之前博客上的相关文章，搬到这里：

## LabVIEW 读取 DLL 函数中的字符串的一个小经验

我们以前在编写 LabVIEW 调用 C DLL 的程序的时候，使用以下策略读取 DLL 函数中的字符串：

1\. 如果能够确保在符串小于 256，比如参数名的字符串等等，就直接在 LabVIEW 中开辟一块大小为 256 的内存。  
2\. 如果不能保证字符串大小，就要为一个读取字符串的 VI 提供两个 DLL 接口函数，先调用一个函数得到字符串的大小，然后在 VI 中开辟内存，之后调用第二个函数得到字符串。

我在使用中，发现第一条策略不保险，应该全部使用第二条策略。

**Reply by ruanqizhen** [reply]
问题（Frank）： 我用 Labwindow 编写了一个读文件的动态库, 即向动态库传递文件路径及文件名和某特定字符串,然后通个三个参数返回读到的值, 在labVIEW里调用该动态库,结果返回值老是显示打开文件失败, 不知错误出现在那里,另外在LabVIEW里如何调试确定传到动态库的参数是符合函数参数格式的呢?该函数在Labwindow里调试没有问题.请大侠指点迷津,不胜感激!

回答：     我好久没有用过 CVI 了，计算机上也没有装，不过用 CVI 来调试，应该和用 VC 来调试原理是相同的，步骤也想类似。我就以 VC 为例说明一下。首先在 Debug 模式下 build 出一个 DLL 来。 （VC 7.1 即便是 release 模式下也可以设置断点，单步运行，但别的编译器不一定行。）然后用这个新的 Debug DLL 覆盖原有的DLL。

关闭 LabVIEW，点击 VC 菜单 Debug->Start (F5)。因为工程生成的是不可以直接执行的 DLL 文件，这是 VC 会弹出一个对话框，问你用什么运行。选择浏览，然后找到 LabVIEW.exe。（这个可执行文件也可以在工程属性中 Debugging->Command 一栏设置。）之后，VC 就会把 LabVIEW 调用起来。

在 VC 中设置好断点。在 LabVIEW 中运行想要调试的 VI。程序会停在你设置断点的地方。

**Reply by ruanqizhen** [reply]
之前博客下的留言：

Qizhen
2009年03月11日 17:25 编辑
这个问题很具体，没没法调试也是找不到问题的。你最好找周围熟悉C语言的人帮忙看一下。

回复
梦蝶
2009年03月3日 20:36 编辑
能帮我看看这段数据采集卡的驱动有没有什么问题啊为什么我编译不了啊#include <windows.h>#include <dos.h>#define DLLExport_declspec(dllexport)DLLExport float hydaq (int chreg){ float value;int i,a,c,base,dl,dh;base=0x280;a=inportb(base+4);outportb(base,0);outportb(base+1,chreg);for(i=0;i<70;i++)outportb(base+2,0);do{c=inportb(base);}while(!(c&0x40));dl=inportb(base+3);dh=inportb(base+4)&0xf;value=((dh*256+dl)-2048)*10/4096;return value;}

回复
Qizhen
2008年05月27日 10:03 编辑
天鹰，你这恐怕是很细致的某一处错误导致的问题，恐怕不是我一下就能说到问题所在的。一般这样常见的问题就是参数配置有误，你还没有理解他的函数每个参数的意义，应当如何设置。建议和厂家或做过类似工作的人联系一下，问他们应当如何配置。

回复
新刚
2008年05月26日 17:38 编辑
长侠:
你好，我看了你的博客，关于labview调用dll的文章，我最近就在调试，我用的是中泰pci8333采集卡，外部使用数字信号发生器产生两路正旋波，用8333来采集，其中就是用到了cln调用8333的动态库pci8kp.dll，但调试的返回值均为0，表明驱动起来了，但就是采不回来数，我在想是不是我的调用语句使用得有问题？还请长侠多多指教！

                                                                                              天鹰

回复
Tom Lu
2007年04月6日 08:42 编辑
我一般是用MessageBox将传进来的参数显示出来看对不对, 虽然很傻的方法，但很实用。

回复
arrowpig
2007年04月4日 12:16 编辑
首先要确定调用规则是不是设对了,LabVIEW里面的配置一定要和被调函数一致.__cdecl还是__stdcall，这两者的压栈顺序一样，从右向左，但是负责清栈不一样。其次要确定LabVIEW里设定的参数个数，类型和被调函数一致。LabVIEW是动态调用DLL，不会在编译时替你检查参数的个数和类型。运行时只是完全按照你的配置压栈，退栈。文件打开失败一般有两个原因，文件名字不对，或者权限有问题。首先debug进你的Dll，确定文件名是不是被正确传递了。如果传递不正确，试着确认一下LabVIEW CLN中的配置，例如，C的string用CSTR。如果文件名字是对的，确定文件I/O操作的返回值。LabVIEW有个特点，LabVIEW的VI和LabVIEW本身是在同一个进程中运行的，也就是说就算VI停止运行，你的DLL还是在内存中，看看是不是你的文件正在被别人打开，而你又试图以独占方式打开文件。
关于调试，Debug和Release的区别在于Debug模式下编译器会试图加入一些附加信息以给调试者更多的提示，设置_DEBUG宏，assert函数生效。Release模式下设置编译器产生调试信息后，在调试时和Debug没有本质区别。在MSVC中可以由DLL工程F5加载LabVIEW进程，或者在Debug菜单下以Attach Process加载。

回复

**Reply by ruanqizhen** [reply]
之前博客下的留言：

Qizhen
2009年05月18日 21:05 编辑
LabVIEW不能回收，若C语言开辟了内存，则必须提供销毁内存的函数，LabVIEW调用该函数销毁其内存。

回复
Kun
2009年05月18日 17:44 编辑
请问在外部C语言中分配的一块内存，labview程序是怎么收回的？

回复
YALU
2008年05月5日 01:28 编辑
最近怎样? 我很久没有上网了! 终于复活了!

回复
竹贤
2008年05月4日 21:13 编辑
请你简要介绍一下这几个VI吧。

回复
mnihj
2008年05月4日 11:31 编辑
这个模块太有用了，以后不用再因为返回指针而费尽心思了。

回复

**Comment by ruanqizhen** [comment]
之前博客上的相关文章，搬到这里：

### 为什么在CLN节点中，会自动配置某些DLL函数的参数信息

这是一个网友问我的问题，我开始也不了解。后来跟同事打听到了一些信息。

有些DLL，比如说是使用LabVIEW生成的DLL，再LabVIEW中，使用“Call Library Function Node”调用DLL中的函数，选择好一个函数，CLN节点自动就把这个函数的参数信息添加上去了。这样，编程者就不需要再对照着头文件去给它配置参数信息。但是，对于绝大多数DLL，比如通过一般步骤，在VC下编译出来的DLL，使用CLN节点选择了函数后，还要手工为其配置参数信息。它们的区别在哪呢？

那些可以识别参数信息的DLL，是因为它们把参数定义的信息，以IDL/ODL文件格式，嵌入到了DLL文件中。LabVIEW，CVI是可以把这些信息嵌进去的。其它编译器也许也可以把信息嵌入DLL中，但是具体如何操作我就没研究过了。

LabVIEW，VB等编程语言可以识别嵌入DLL的参数设置信息，在这些语言下使用这种使用有参数信息的DLL，更加便捷。

**Reply by ruanqizhen** [reply]
之前博客下的留言：

Jing
2008年09月28日 09:40 编辑
嗯，差不多了。

回复
arrowpig
2008年09月26日 22:25 编辑
小牛牛，你的NET已经做好了？

回复
Jing
2008年09月26日 17:36 编辑
在VC中的Link属性中可以选择IDL文件插入。
建议使用.NET的dll（assembly），不需要配置参数。

回复
